File: mergesort.py
#!/usr/bin/python3

from typing import List


def merge(left: List[int], right: List[int]) -> (List[int], int):
    """
    Merge two arrays of ints.
    :return: Merged (sorted) array of ints and number of inversions
    """
    results = []
    inversions = 0
    i = 0
    j = 0
    count = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            results.append(left[i])
            i += 1
        else:
            results.append(right[j])
            j += 1
            inversions += len(left) - i
    results += left[i:] + right[j:]
    return results, inversions


def merge_sort(array: List[int]) -> (List[int], int):
    """
    Use mergesort algorithm to sort array.

    :return: Return sorted array of ints and number of inversions.
    """
    n = len(array)
    if n < 2:
        return array, 0
    mid = n // 2
    left = array[:mid]
    right = array[mid:]
    left, linv = merge_sort(left)
    right, rinv = merge_sort(right)
    results, inv = merge(left, right)
    return results, inv + linv + rinv


if __name__ == '__main__':
    input = {
        0:  [1, 1, 1, 2, 2],
        4:  [2, 1, 3, 1, 2],
        15: [5, 4, 3, 2, 1, 0],
    }

    for ans in input.keys():
        arr = input[ans]
        n = len(arr)

        arr = merge_sort(arr)
        print(arr)

File: mergesort.go
package main

import "fmt"

// Merge two arrays, returning sorted array and number of inversions.
func merge(l, r []int) ([]int, int) {
	var srtd []int
	var inv int
	i, j := 0, 0
	for i < len(l) && j < len(r) {
		if l[i] <= r[j] {
			srtd = append(srtd, l[i])
			i += 1
		} else {
			srtd = append(srtd, r[j])
			j += 1
			inv += len(l) - i
		}
	}
	srtd = append(srtd, l[i:]...)
	srtd = append(srtd, r[j:]...)
	return srtd, inv
}

// Use mergesort to get the sorted array and number of inversions.
func MergeSort(arr []int) ([]int, int) {
	if len(arr) < 2 {
		return arr, 0
	}
	l, li := MergeSort(arr[:len(arr)/2])
	r, ri := MergeSort(arr[len(arr)/2:])
	rv, ti := merge(l, r)
	return rv, li + ri + ti
}

func main() {
	a := []int{2, 4, 1, 0}
	sorted, inversions := MergeSort(a)
	fmt.Println(sorted)
	fmt.Println(inversions)
}
